language-plutus-core/plutus-ir/Language/PlutusIR/TypeCheck/Internal.hs

```diff
@@ -0,0 +1,615 @@
+inferTypeM (Unwrap ann term)        = do
+    vTermTy <- inferTypeM term
+    case unNormalized vTermTy of
+        TyIFix _ vPat vArg -> do
+            k <- inferKindM $ ann <$ vArg
+            -- Subparts of a normalized type, so normalized.
+            unfoldFixOf (Normalized vPat) (Normalized vArg) k
+        _                  -> throwError (TypeMismatch ann (void term) (TyIFix () dummyType dummyType) vTermTy)
+-- [check| G !- ty :: *]    ty ~>? vTy
+-- -----------------------------------
+-- [infer| G !- error ty : vTy]
+inferTypeM (Error ann ty)           = do
+    checkKindM ann ty $ Type ()
+    normalizeTypeM $ void ty
+
+inferTypeM (Let ann1 recurs bs inTerm) = do
+    tyInTerm <- case recurs of
+        -- [check| G !- ty :: *]  [infer| G,withBindsLinear(b) !- inT : ty] (forall b in bs. [checkWellFormed| Glinear !- b]) 
+        -- -------------------------------------------------
+        -- [infer| G !- (let (nonrec) bs in inT) :: ty
+        NonRec ->
+            foldr
+               (\b acc -> do
+                   checkWellformBind NonRec b
+                   -- linear scope: the new identifiers will be visible in the next binding
+                   withBind b acc)
+               (inferTypeM inTerm)
+               bs
+        -- [check| G !- ty :: *]  G'=G,withBinds(bs)  [infer| G' !- inT : ty] (forall b in bs. [checkWellFormed| G' !- b])
+        -- -------------------------------------------------
+        -- [infer| G !- (let (rec) bs in inT) :: ty
+        Rec -> do
+            -- Check that there are no conflicting new identifiers (same-name)
+            -- that are introduced by the bindings of this the letrec
+            -- separately check the VarTypes environment for conflicts
+            -- TODO: move these 2 checks to a different pass, e.g. in renaming phase
+            assertNoRecConflict $ foldMap bindingNewIds bs
+            -- separately check the VarkInds environment for conflicts
+            assertNoRecConflict . mapMaybe bindingNewTyId $ toList bs
+            -- now with all bindings in the env, check for wellformedness
+            withBinds bs $ do
+              checkWellformBinds bs
+              inferTypeM inTerm
+    -- In paper's Fig.6, T-Let and T-LetRec rules dictate that: G !- inTerm :: *
+    -- Durining implemenetation, however, this leads to scoping errors because
+    -- the G env is not augmented to see the kinds introduced by any (nested) let(s)
+    -- So we disable this check: `checkKindM ann (ann <$ unNormalized tyInTerm) $ Type ()`
+    -- Furthermore, this check would be unnecessary afterall, because of the proof-by-construction:
+    -- All terms that can be constructed have types that are *-kinded.
+    pure tyInTerm
+
+  where
+    withBinds :: NonEmpty (Binding TyName Name uni ann) -> TypeCheckM uni ann a -> TypeCheckM uni ann a
+    withBinds = flip . foldr $ withBind
+
+    checkWellformBinds :: NonEmpty (Binding TyName Name uni ann) -> TypeCheckM uni ann ()
```

Comment:
I'd inline this - it's only used once and this actually conceals the fact that you're passing `Rec` which is important.